docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,672 @@
+    - [PushNotificationPreferences](#pushnotificationpreferences)
+    - [PushNotificationDeviceToken](#pushnotificationdevicetoken)
+    - [PushNotificationOptions](#pushnotificationoptions)
+    - [PushNotificationPreferencesAcknowledgment](#pushnotificationpreferencesacknowledgement)
+    - [ContactCodeAdvertisement](#contactcodeadvertisement)
+    - [PushNotificationAdvertisementInfo](#pushnotificationadvertisementinfo)
+    - [PushNotificationQuery](#pushnotificationquery)
+    - [PushNotificationQueryInfo](#pushnotificationqueryinfo)
+    - [PushNotificationQueryResponse](#pushnotificationqueryresponse)
+    - [PushNotification](#pushnotification)
+    - [PushNotificationRequest](#pushnotificationrequest)
+    - [PushNotificationAcknowledgement](#pushnotificationacknowledgement)
+  - [Partitioned topic](#partitioned-topic)
+  - [FAQ](#faq)
+  - [Changelog](#changelog)
+    - [Version 0.1](#version-01)
+  - [Copyright](#copyright)
+
+## Reason
+
+Push notifications for IOS devices and for some android device can only be
+implemented relying on [APN service](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1) for IOS or [Firebase](https://firebase.google.com/) for 
+Android devices that do not support background services or that often kill the 
+background service.
+Therefore Status provides a set of Push notification services that can be used
+to achieve this functionality.
+
+Because this can't be safely implemented in a privacy preserving manner, clients
+MUST be given an option not to enable push notifications and not to send push notifications.
+
+## Requirements
+
+Status needs to run their own [gorush](https://github.com/appleboy/gorush) 
+publicly accessible server for sending the actual notification. This cannot be community run as it requires sharing private credentials that only status has access to.
+
+## Registering with the push notification service
+
+A client MAY register with a Push Notification service of their choice.
+
+[//]: # (We can have a separate partitioned topic scheme, or individual based on PK)
+
+A message MUST be sent to the [partitioned topic](#partitioned-topic) for the
+public key of the node, encrypted with this key.
+
+The content of the message MUST contain the following [protobuf record](https://developers.google.com/protocol-buffers/):
+
+```protobuf
+
+message PushNotificationOptions {
+  boolean enabled = 1;
+  repeated string allowed_user_list = 2;
+  repeated string block_user_list = 3;
+  repeated string block_chat_list = 4;
+}
+
+message PushNotificationDeviceToken {
+  enum TokenType {
+    UNKWOWN_TOKEN_TYPE = 0;
+    APS_TOKEN = 1;
+    FIREBASE_TOKEN = 2;
+  }
+  TokenType token_type = 1;
+  string token = 2;
+  string installation_id = 3;
+  PushNotificationPreferences preferences = 4;
+}
+
+message PushNotificationPreferences {
+  repeated DeviceToken device_tokens = 1;
```

Comment:
Outside of DeviceToken, is the unique identifier here DeviceToken+ID? How do we deal with the 1-device-N-user case here?

Reply:
> 1-device-N-user

Not sure I understand exactly the question, you mean multiple users using the same device? Which is essentially, how do we handle multiple accounts on a single installation?

If that's the case, is `installation_id`. Installation id is different for each generated account, so if you have 3 accounts on your device, you will have 3 different installation ids. this is passed to the push notification, so on opening the notification, we can already present the user with the right account and it can type the password.

hope I understood the question correctly.

Reply:
Multiple accounts/users on one device, correct.